### PR TITLE
Fix edge preview props are not matched with cookie

### LINKED ISSUE #52080: Draft Mode isEnabled is always false on the Edge Runtime
### Verify canary release

- [X] I verified that the issue exists in the latest Next.js canary release

### Provide environment information

```bash
Operating System:
      Platform: linux
      Arch: x64
      Version: #1 SMP PREEMPT_DYNAMIC Thu Jun 15 02:15:40 UTC 2023
    Binaries:
      Node: 18.16.0
      npm: 8.11.0
      Yarn: 1.22.19
      pnpm: N/A
    Relevant Packages:
      next: 13.4.8-canary.14
      eslint-config-next: N/A
      react: 18.2.0
      react-dom: 18.2.0
      typescript: 4.9.5
    Next.js Config:
      output: N/A
```

### Which area(s) of Next.js are affected? (leave empty if unsure)

App directory (appDir: true)

### Link to the code that reproduces this issue or a replay of the bug

https://github.com/alvarlagerlof/next-js-draft-mode-repro

### To Reproduce

Run the repro in dev or build.

1. Click the link on / to enable draft mode.
2. Go back to /
3. Draft mode is not enabled

### Describe the Bug

`isEnabled` from `draftMode()` always returns false with `export const runtime = "edge"`

### Expected Behavior

`isEnabled` returns `true` after `draftMode().enable()` is ran,

### Which browser are you using? (if relevant)

_No response_

### How are you deploying your application? (if relevant)

Seems to happen locally, in dev, build, and on Vercel

<sub>[NEXT-1477](https://linear.app/vercel/issue/NEXT-1477/draft-mode-isenabled-is-always-false-on-the-edge-runtime)</sub>

### LINKED ISSUE #67075: Draft Mode does not work in Middleware
### Link to the code that reproduces this issue

https://github.com/JKarlavige/nextjs-draftmode-reproduction

### To Reproduce

1. Clone reproduction repository locally
2. Install packages with `npm i`
3. Run app with `npm run dev`
4. In a browser, hit the following url: `localhost:3000/api/draft?secret=secret-token&slug=preview-page`

### Current vs. Expected behavior

Note - Draft mode on the page itself works correctly. `draftMode().isEnabled` returns true. 

Current behavior:
In middleware, `draftMode().isEnabled` returns `false` after hitting the draft endpoint in step 4 above.

Expected behavior:
`draftMode().isEnabled` should return `true` after the draft endpoint has been hit.

### Provide environment information

```bash
Operating System:
  Platform: darwin
  Arch: arm64
  Version: Darwin Kernel Version 23.5.0: Wed May  1 20:12:58 PDT 2024; root:xnu-10063.121.3~5/RELEASE_ARM64_T6000
  Available memory (MB): 16384
  Available CPU cores: 10
Binaries:
  Node: 20.11.1
  npm: 10.2.4
Relevant Packages:
  next: 14.2.4 // Latest available version is detected (14.2.4).
  react: 18.3.1
  react-dom: 18.3.1
```

### Which area(s) are affected? (Select all that apply)

Middleware

### Which stage(s) are affected? (Select all that apply)

next dev (local), Vercel (Deployed)

### Additional context

On `next` version  14.1.4 (which we currently use in production), we experience inconsistencies where draft mode in middleware sometimes returns `true`, and on the following middleware run it returns `false`. 

On version 14.2.4, it appears that draft mode always returns `false`.
middleware-draftmode

### PR DESCRIPTION
**Pull request by huozhi**
### What

Chnage the way of preview props injection to edge runtime, directly read from env vars of preview props instead of writing to and reading from prerender manifest.js

### Why

Previously we're trying to make these preview props values become deterministic that we can replace in edge deployment pipeline in #64521 

But the way of serializing process env vars in edge runtime is not correct. They'll remain as string "process.env.xxx" in the manifest and also after consumed. This PR fixes that behavior, instead of writing it into manifest, alwyas consuming from process.env.var directly.

I created a shared util to access the preview props of edge runtime across all the templates.

On draft provider side, we still need to handle dev mode case when preview id is `development-id`, but we already have the cookie, it cannot be aligned with the preview id. So we do a fallback check for dev mode if the cookie is present and preview id is `development-id` then we still treat it as draft mode is enabled.

Fixes #52080
Fixes #67075

Closes NEXT-3541

### CODE REVIEW

**Comment by huozhi** [comment]
* **#67779** <a href="https://app.graphite.dev/github/pr/vercel/next.js/67779?utm_source=stack-comment-icon" target="_blank">Graphite</a> 👈
* `canary`

This stack of pull requests is managed by Graphite. <a href="https://stacking.dev/?utm_source=stack-comment">Learn more about stacking.</a>
<h2></h2>

 Join @huozhi and the rest of your teammates on <a href="https://graphite.dev?utm-source=stack-comment">Graphite <b>Graphite</b></a>

**Comment by ijjk** [comment]
## Failing test suites 
Commit: 3aac3c868a7302c0710c148456fac968a6ac2274

`TURBOPACK=1 pnpm test test/integration/react-streaming/test/index.test.js` (turbopack)
- streaming dev dev > should support streaming for fizz response
- streaming dev dev > should not stream to crawlers or google pagerender bot
- streaming dev dev > should render 500 error correctly
- streaming dev dev > should render fallback if error raised from suspense during streaming

<details>
<summary>Expand output</summary>

  ● streaming dev dev › should support streaming for fizz response

    expect(received).toBe(expected) // Object.is equality

    Expected: true
    Received: false

      43 |           // Streaming is disabled for pages, no fallback should be rendered.
      44 |           expect(gotFallback).toBe(false)
    > 45 |           expect(gotData).toBe(true)
         |                           ^
      46 |         }
      47 |       )
      48 |

      at toBe (integration/react-streaming/test/streaming.js:45:27)
      at testStreamingResponse (integration/react-streaming/test/streaming.js:28:7)
      at Object.<anonymous> (integration/react-streaming/test/streaming.js:55:5)

  ● streaming dev dev › should not stream to crawlers or google pagerender bot

    expect(received).toBe(expected) // Object.is equality

    Expected: "next_streaming_data"
    Received: ""

      88 |     const body = await getNodeBySelector(html, '#__next')
      89 |     // Resolve data instead of fallback
    > 90 |     expect(body.text()).toBe('next_streaming_data')
         |                         ^
      91 |
      92 |     expect(res1.headers.get('etag')).toBeDefined()
      93 |     expect(res2.headers.get('etag')).toBeDefined()

      at Object.toBe (integration/react-streaming/test/streaming.js:90:25)

  ● streaming dev dev › should render 500 error correctly

    expect(received).toContain(expected) // indexOf

    Expected substring: "Error: oops"
    Received string:    "Internal Server Error"

      100 |       if (env === 'dev') {
      101 |         // In development mode it should show the error popup.
    > 102 |         expect(html).toContain('Error: oops')
          |                      ^
      103 |       } else {
      104 |         expect(html).toContain('custom-500-page')
      105 |       }

      at toContain (integration/react-streaming/test/streaming.js:102:22)
          at async Promise.all (index 1)
      at Object.<anonymous> (integration/react-streaming/test/streaming.js:107:5)

  ● streaming dev dev › should render fallback if error raised from suspense during streaming

    expect(received).toContain(expected) // indexOf

    Expected substring: "error-fallback"
    Received string:    "Internal Server Error"

      110 |   it('should render fallback if error raised from suspense during streaming', async () => {
      111 |     const html = await renderViaHTTP(context.appPort, '/err/suspense')
    > 112 |     expect(html).toContain('error-fallback')
          |                  ^
      113 |   })
      114 | }
      115 |

      at Object.toContain (integration/react-streaming/test/streaming.js:112:18)

</details>


Read more about building and testing Next.js in [contributing.md](https://github.com/vercel/next.js/blob/canary/contributing.md).

File: packages/next/src/build/index.ts [hunk 1]
```diff
@@ -357,12 +357,10 @@ async function writeEdgePartialPrerenderManifest(
       dynamicRoutes: {},
       notFoundRoutes: [],
       version: manifest.version,
-      preview: {
-        previewModeId: 'process.env.__NEXT_PREVIEW_MODE_ID',
-        previewModeSigningKey: 'process.env.__NEXT_PREVIEW_MODE_SIGNING_KEY',
-        previewModeEncryptionKey:
-          'process.env.__NEXT_PREVIEW_MODE_ENCRYPTION_KEY',
-      },
+      // Preview props are inlined in the code with dynamic env vars,
```
Comment (ijjk) [inline packages/next/src/build/index.ts:360]: These values were the main reason this manifest was added iirc so should we just fully remove it? 

Could be in a separate PR just noting for cleanup purposes.

File: test/e2e/app-dir/draft-mode-middleware/middleware.ts [hunk 2]
```diff
@@ -0,0 +1,17 @@
+import { NextResponse, type NextRequest } from 'next/server'
+import { draftMode } from 'next/headers'
+
+export function middleware(req: NextRequest) {
+  const { isEnabled } = draftMode()
+
+  // This should return `true` when draft mode enabled
+  // but always returns false.
```
Comment (ijjk) [inline test/e2e/app-dir/draft-mode-middleware/middleware.ts:8]:
```suggestion
```
This is contradictory to above, is it leftover?

**Review by ijjk — APPROVED** [review]
This makes sense to me, left couple suggestions

File: packages/next/src/build/index.ts (hunk 1 above)
Comment (huozhi) [reply packages/next/src/build/index.ts:360]: Sounds good, I'll get a follow up PR to remove it from edge runtime

**Pull request merged**